great_expectations/rule_based_profiler/rule_based_profiler.py

```diff
@@ -295,27 +343,250 @@ def run(
+            for rule_name, rule_config in override_rule_configs.items()
+        }
+        effective_rules.update(override_rules)
 
-        if pretty_print:
-            print(f"\nRuleBasedProfiler class name: {self.name}")
+        return list(effective_rules.values())
 
-            if not self._variables:
-                print(
-                    'Your current RuleBasedProfiler configuration has an empty "variables" attribute. \
-                    Please ensure you populate it if you\'d like to reference values in your "rules" attribute.'
+    @staticmethod
+    def _reconcile_rule(
+        existing_rules: Dict[str, Rule], rule_name: str, rule_config: dict
+    ) -> Dict[str, Any]:
+        effective_rule_config: Dict[str, Any]
+        if rule_name in existing_rules:
+            rule: Rule = existing_rules[rule_name]
+            domain_builder_config: dict = rule_config.get("domain_builder", {})
+            effective_domain_builder_config: dict = (
+                RuleBasedProfiler._reconcile_domain_builder_config(
+                    rule=rule,
+                    domain_builder_config=domain_builder_config,
                 )
+            )
+            effective_parameter_builder_configs: Optional[
+                List[dict]
+            ] = RuleBasedProfiler._reconcile_parameter_builder_configs_for_rule(
+                rule=rule,
+                rule_config=rule_config,
+            )
+
+            effective_expectation_configuration_builder_configs: List[
+                dict
+            ] = RuleBasedProfiler._reconcile_expectation_configuration_builder_configs_for_rule(
+                rule=rule,
+                rule_config=rule_config,
+            )
+            effective_rule_config = {
+                "domain_builder": effective_domain_builder_config,
+                "parameter_builders": effective_parameter_builder_configs,
+                "expectation_configuration_builders": effective_expectation_configuration_builder_configs,
+            }
+        else:
+            effective_rule_config = rule_config
+
+        return effective_rule_config
+
+    @staticmethod
+    def _reconcile_domain_builder_config(
+        rule: Rule,
+        domain_builder_config: dict,
+    ) -> dict:
+        effective_domain_builder_config: dict = {}
+        batch_request: Optional[
+            Union[BatchRequest, RuntimeBatchRequest, dict]
+        ] = domain_builder_config.pop("batch_request", None)
+        if batch_request is None:
+            batch_request = get_batch_request_as_dict(
+                batch_request=rule.domain_builder.batch_request
+            )
+
+        effective_domain_builder_config["batch_request"] = batch_request
+
+        key: str
+        value: Any
+        current_value: Any
+        for key, value in domain_builder_config:
+            if hasattr(rule, f"{key}"):
+                current_value = getattr(rule, key)
```

Comment:
Can we circumvent the need for `hasattr`? I might ask to connect on this live.

Reply:
@cdkini Good question -- the only way would be to require each class to declare its own fields and implement `.to_dict()` -- while this made sense for `BatchRequestBase` subclasses, which we control (and there are only two of them), it is not so good for the `DomainBuilder`, `ParameterBuilder`, and `ExpectationConfiguration` builder subclasses, because we do not control them (users will develop their custom versions, and there will be many of them).  Do you think there is a better way?  If nothing immediate comes to mind, then this generic approach will be good for now.

Reply:
Can we convert the `Rule` to a dict? And then iterate over keys?

I'm really reluctant to use `getattr/setattr` so maybe let's talk about this.